## Data Analsyis: Cars

Welcome, coders! You've chosen to analyze the car data set. We will be looking at the data contained in the cars.csv file, which you can check out on the left-hand side of the screen - just make your way back here when you're done.

This 2022-2023 data set contains information on over 2500 different cars, including make, model, price, horsepower, fuel efficiency and more. We're going to look at the price of different cars based on their make.

First things first, we're going to have to get Python to read the file and put it into something we can work with - like a list!

Run the kernel below to do this for you.

In [36]:
# First we're going to import our helpful Python libraries

import numpy as np
import csv
from matplotlib import pyplot as plt

# We start by creating a blank list called master that we will be able to add to as the csv reader goes through the file
master = []

# Then we open the file using the csv reader and use next() to skip over the header line

with open('cars.csv', newline='') as csvfile:
    file = csv.reader(csvfile)
    next(file)
    
# Now we use a for loop to iterate through the file.  This reads each row as a separate item.

    for row in file:

# We use the append method on the master list to add each row as a list item as we iterate through the file.
        
        master.append(row)
        

If we look at our .csv file, we can see that 'Brand' is the second column and 'Price' is the fourth. Let's try printing just the fourth and fifth indices of the master list.

In the kernel below, use the print function to display the fourth and fifth indices of master. (Remember: indexing starts at 0!)

In [38]:
# Enter code here


If you got something that starts with ['4', 'Honda', 'HR-V VTi-S 2022'... you did it right! But... that's not what we're looking for.

When we added all of the data to the list, it added each row of the csv file as an item. However, each row also had 5 columns of data, so what we've got now is a bunch of lists within a list. This is ridiculist! (Sorry)

What we're looking at now is no longer a flat, one-dimensional list, but a list with two dimensions (kind of like the table we saw the data in). But fret not, Python can help us with that.

Numpy's shape function can tell us the dimensions that we're working with. Run the kernel below to determine the shape of our master list using the shape function.

In [39]:
print(np.shape(master))

(2602, 76)


Hm, so this tells us that there are 2602 list elements (one for each car), each with 76 elements of their own (one for each different type of data: name, make, price, etc). We want only the second and fourth item in each of the 182 elements. We can use a for loop to cycle through each element and put only the items we do want into a new list.

To reference a range of items, rather than adding them separately, I can use a colon in between two indices. For example,

my_list.append(row[1:5])

would add the elements with indices 1 through 4 (notice the 5 in the brackets means up to, but not including index 5) from the row to my_list.

I can also use a "step" by adding another colon. For example,

my_list.append(row[1:5:2])

would add every second elements with indices 1 through 4 - i.e. it would add my list[1], skip over my_list [2], add my list[3], and skip over my_list[4].

The for loop is set up for you in the kernel below, but is missing the indices inside the square brackets. Fix the code to include the correct indices. Then, print just the first element of new_master. You may move on to the next step when you successfully produce an output of ['Honda', '31670']

In [41]:
new_master = []
for row in master:
    new_master.append(row[1:4:2])

print(new_master[0])

['Honda', '31670']


Great! Now, we can get to work. Since we want look at prices by brand, we're going to create separate lists for each brand with their prices. We will do this using another for loop to iterate through our new master list. 

The loop is created for you in the kernel below for the brands Honda, Ford, Jeep, Toyota, and Mazda, however it's missing the indices. Use the clues to fill in the missing code. 
Then, use the len() function to check the lengths of each of the genre lists. You may proceed when you get the following lengths:

60 60 60 60 60

Remember: This new master list has 2602 "columns" and only **2** "rows" now.

In [44]:
honda = []
ford = []
jeep = []
toyota = []
mazda = []

a = # Insert Brand index here
b = # Insert Price index here

for row in master:
    if "Honda" in row[a]:
        honda.append(float(row[b]))
    elif "Ford" in row[a]:
        ford.append(float(row[b]))
    elif "Jeep" in row[a]:
        jeep.append(float(row[b]))
    elif "Toyota" in row[a]:
        toyota.append(float(row[b]))
    elif "Mazda" in row[a]:
        mazda.append(float(row[b]))

60 60 60 60 60
[31670.0, 28000.0, 25000.0, 27800.0, 26900.0, 25440.0, 29990.0, 65000.0, 46230.0, 33760.0, 31070.0, 33950.0, 32450.0, 40000.0, 36830.0, 37870.0, 32790.0, 31610.0, 41210.0, 37340.0, 28900.0, 25760.0, 33250.0, 28000.0, 18030.0, 16150.0, 16330.0, 14340.0, 16080.0, 14330.0, 31800.0, 28690.0, 26890.0, 23650.0, 23650.0, 30690.0, 28890.0, 24850.0, 27500.0, 28580.0, 15520.0, 18960.0, 44730.0, 16190.0, 21870.0, 41780.0, 38800.0, 39970.0, 27450.0, 32350.0, 26570.0, 25070.0, 23650.0, 26890.0, 25320.0, 44310.0, 23650.0, 34640.0, 33860.0, 33260.0] [36475.0, 59474.0, 74474.0, 51974.0, 65970.0, 38560.0, 30000.0, 109140.0, 59940.0, 45760.0, 55905.0, 79620.0, 77170.0, 67115.0, 50000.0, 55000.0, 35960.0, 28845.0, 29840.0, 34040.0, 38460.0, 52905.0, 51990.0, 47990.0, 36590.0, 27950.0, 24450.0, 505750.0, 502300.0, 500000.0, 349000.0, 68380.0, 70705.0, 34965.0, 52190.0, 58500.0, 47780.0, 46640.0, 39340.0, 37940.0, 43100.0, 27950.0, 34900.0, 58830.0, 45760.0, 50600.0, 44560.0, 45850.0, 61830.

Awesome! Now let's do a statistical analysis. Use Numpy's quantile function and mean function to print the five-number summaries and means for each brand. Move on to the next step when your output looks like this:

Honda:

Min: *number*, Q1: *number*, Median: *number*, Q3: *number*, Max: *number*, Mean: *number*

Ford: 

Min: *number*, Q1: *number*, Median: *number*, Q3: *number*, Max: *number*, Mean: *number*

and so on...

In [45]:
# Enter code here

Great! Now let's plot these all together and see if they match our statistical summaries. 

In the kernel below, first create a new list that has each of the genre lists as an element.

Example: new_list = [brand1, brand2, ...]

Then, create a list called labels with a string for each genre.

Example: labels = ['Brand 1', 'Brand 2', ...]

Then, use Pyplot's boxplot function to plot this new list. Include labels=labels as an input to display the genres on the plot. Use Pyplot's title, xlabel, and ylabel functions to label the graph.

In [46]:
# Enter code here

Fabulous! You may notice some circles on the graph - these are outliers. Try editing your code above to add showfliers=False as an optional input in the boxplot function and re-run the kernel.

Wonderful! Now it's your turn to play around with the data. The csv file contains information on many more brands, including:

Honda,
BMW,
Lexus,
Hyundai,
Toyota,
KIA,
Nissan,
Audi,
Chevrolet,
Ford,
Mercedes,
Porsche,
Infiniti,
Jaguar,
Cadillac,
Land Rover,
Jeep,
Volkswagen,
Maserati,
Subaru,
Dodge,
Mazda,
Chrysler,
Aston Martin,
Ferrari,
Lamborghini,
Bugatti,
Bentley,
Rolls Royce,
Mclaren,
Lincoln,
Alfa Romeo,
Volvo,
MINI,
Fiat,
Acura,
Genesis,
Buick,
GMC,
and Tesla

I've copy-pasted one of the previous kernels into the kernel below. Try editing this code to extract data on different brands (any brands you'd like), then create a new box plot comparing those.

You may also play around with the customization using the example in the introduction as a guide.

When you're happy with your boxplot, call me over to check your work.

Happy coding!

In [ ]:
honda = []
ford = []
jeep = []
toyota = []
mazda = []

a = # Insert Brand index here
b = # Insert Price index here

for row in master:
    if "Honda" in row[a]:
        honda.append(float(row[b]))
    elif "Ford" in row[a]:
        ford.append(float(row[b]))
    elif "Jeep" in row[a]:
        jeep.append(float(row[b]))
    elif "Toyota" in row[a]:
        toyota.append(float(row[b]))
    elif "Mazda" in row[a]:
        mazda.append(float(row[b]))